### DISCUSSION TITLE
How to speed up binary transfer?

### DISCUSSION THREAD

**Discussion by adamplonka** [AArnott/Nerdbank.Streams#503] [Q&A]
I'm trying to send a large binary file using `MultiplexingStream` over `TcpSocket`. I tried to send it directly using sockets (for now on loopback interface) and I'm able to transfer over 10GB/s.
When I try to do the same using `MultiplexingStream` and return `Stream`, `PipeReader` or create separate channel I could get up to only 80MB/s.

I tried to increase the buffer sizes but with no luck. Does the `MultiplexingStream` add so much overhead that it's not possible to get it any faster?

// after lot of tries I found out that there is a significant difference in speed between versions and I tried to use the newest one
Average speeds I got on my laptop are:
V1 - 118.11 MBps
V2 - 41.5 MBps
V3 - 48.0 MBps

Anyway - this application will run on a much slower device which during my initial tests got only up to 8MBps on V3; there's still a huge gap between multiplexed stream and raw stream speeds. Is there any way to make it faster?
Great library btw

**Comment by AArnott** [comment]
Thanks for bringing this up. The multiplexing protocol was designed to be lightweight and minimal. I'm very surprised to hear of this speed decrease of this magnitude.  I'll take a look.

**Reply by AArnott** [reply]
I'm now tracking this with #505
